simulator.py

```diff
@@ -0,0 +1,169 @@
+    """
+    def __init__(self, region, instruments):
+        self.quantity = collections.namedtuple('value', ['value', 'quantity'])
+        self.cur_asset = {
+            instrument : self.quantity(0, 0) for instrument in instruments}
+        self.cur_asset[region] = self.quantity(1, 1)
+        self.__portfolio = list()
+        self.__pnl = Data('pnl', ['pnl', 'trading_cost'])
+
+    def get_results(self):
+        """
+        Return results.
+        """
+        result = Data('result', ['pnl', 'trading_cost', 'position'])
+        for index in range(0, len(self.__pnl), 1):
+            result.append(
+                (self.__pnl[index].date, self.__pnl[index].pnl,
+                 self.__pnl[index].trading_cost, self.__portfolio[index]))
+        return result
+
+    def produce_portfolio(self, today, alpha_formula, data):
+        """
+        Find daily portfolio.
+        """
+        portfolio = alpha_formula(today, data)
+        portfolio.normalize()
+        self.__portfolio.append(portfolio)
+
+    def produce_pnl(self, region, cost_function, target_data):
+        """
+        Calculate pnl for alphas.
+        This method will be called by simulation day by day while inputting
+        a new portfolio.
+
+        While calculating pnl,
+        we should take a look at our asset and trading cost.
+            (1) Total asset
+                We assume that our assets are the same everyday (equal to
+                self.initial_asset). To achieve this goal, we open a virtual
+                account to borrow/lend money to the investing account (this
+                account is self.cur_asset[self.base]).
+            (2) Trading cost
+                There are different functions about calculating trading cost
+                in different investing targets.
+                This method should import different modules while targets are
+                different.
+
+                In our template function, we calculate trading cost while
+                selling out the asset.
+                The trading_cost function could be built by alpha writers.
+                The API will adjust in the future, while the fuction needs
+                more data.
+        """
+        today = self.__portfolio[-1]
+        cost = 0
+        trading_value = 0
+        profit = 0
+        for instrument, proportion in today.items():
+            if target_data[instrument][-1].sell is not None:
+                now = self.quantity(
+                    proportion,
+                    proportion / target_data[instrument][-1].sell
+                )
+                trading_value += now.value
+                profit += (
+                    (self.cur_asset[instrument].quantity - now.quantity) *
+                    target_data[instrument][-1].sell)
+                cost += cost_function(
+                    self.cur_asset[instrument],
+                    now,
+                    target_data[instrument][-1]
+                )
```

Comment:
While profit is related to cost, it makes more sense to handle them together.
That is, we can calculate profit and cost in the cost function, which offers more flexibility to more complex cost function.